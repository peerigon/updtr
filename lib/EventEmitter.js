"use strict";

var events = require("events");

var EventEmitter = events;

if (typeof EventEmitter !== "function") {
    EventEmitter = events.EventEmitter;
}

module.exports = EventEmitter;
