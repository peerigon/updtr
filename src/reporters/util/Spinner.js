import cliSpinners from "cli-spinners";

const isWin = process.platform === "win32";
const winFallback = "simpleDotsScrolling";

export default class Spinner {
    constructor(spinnerName) {
        const spinner = cliSpinners[isWin === true ? winFallback : spinnerName];

        this.frames = spinner.frames;
        this.interval = spinner.interval;
        this.length = Math.max(...this.frames.map(frame => frame.length));
    }
    valueOf() {
        const currentInterval = Math.floor(Date.now() / this.interval);
        const currentFrame = currentInterval % this.frames.length;

        return this.frames[currentFrame];
    }
    toString() {
        return this.valueOf();
    }
}
