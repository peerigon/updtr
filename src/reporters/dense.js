import chalk from "chalk";
import unicons from "unicons";
import Projector from "./util/Projector";
import Terminal from "./util/Terminal";
import Message from "./util/Message";
import Spinner from "./util/Spinner";
import Indicator, {
    INDICATOR_NEUTRAL,
    INDICATOR_PENDING,
    INDICATOR_OK,
    INDICATOR_FAIL,
} from "./util/Indicator";
import customConfigToLines from "./util/customConfigToLines";
import pluralize from "./util/pluralize";
import handleError from "./util/handleError";
import msToString from "./util/msToString";
import {
    filterSuccessfulUpdates,
    filterFailedUpdates,
} from "../tasks/util/filterUpdateResults";

const spinner = new Spinner("dots");

function updatingLine(updateTask) {
    return [
        new Indicator(INDICATOR_PENDING),
        chalk.bold(updateTask.name),
        chalk.grey("updating"),
        updateTask.rollbackTo,
        chalk.grey(unicons.arrowRight),
        updateTask.updateTo + chalk.grey("..."),
    ].join(" ");
}

function testingLine(updateTask) {
    return [
        new Indicator(INDICATOR_PENDING),
        chalk.bold(updateTask.name),
        chalk.grey("testing..."),
    ].join(" ");
}

function rollbackLine(updateTask) {
    return [
        new Indicator(INDICATOR_FAIL),
        chalk.bold.red(updateTask.name),
        chalk.grey("rolling back"),
        updateTask.updateTo,
        chalk.grey(unicons.arrowRight),
        updateTask.rollbackTo + chalk.grey("..."),
    ].join(" ");
}

function successLine(updateTask) {
    return [
        new Indicator(INDICATOR_OK),
        chalk.bold(updateTask.name),
        updateTask.updateTo,
        chalk.grey("success"),
    ].join(" ");
}

function failLine(updateTask) {
    return [
        new Indicator(INDICATOR_FAIL),
        chalk.bold.red(updateTask.name),
        updateTask.updateTo,
        chalk.grey("failed"),
    ].join(" ");
}

function excludedLine(excluded) {
    return [
        new Indicator(INDICATOR_NEUTRAL),
        chalk.bold(excluded.name),
        chalk.grey(excluded.reason),
    ].join(" ");
}

function cmdToLines(description, cmd) {
    const lines = Array.isArray(description) === true ?
        description :
        [description];

    return lines.concat([chalk.grey(`> ${ cmd } `), spinner]);
}

export default function (updtr, reporterConfig) {
    const terminal = new Terminal(reporterConfig.stream);
    const projector = new Projector(terminal);
    const startTime = Date.now();
    let excludedModules;

    updtr.on("start", ({ config }) => {
        terminal.append(customConfigToLines(config));
    });
    updtr.on("init/install-missing", ({ cmd }) => {
        projector.display(
            cmdToLines(
                "Installing missing dependencies" + chalk.grey("..."),
                cmd
            )
        );
    });
    updtr.on("init/collect", ({ cmd }) => {
        projector.display(
            cmdToLines("Looking for outdated modules" + chalk.grey("..."), cmd)
        );
    });
    updtr.on("init/end", ({ updateTasks, excluded }) => {
        excludedModules = excluded;
        projector.stop();
        if (updateTasks.length === 0 && excluded.length === 0) {
            terminal.append(["Everything " + chalk.bold("up-to-date")]);
        } else if (updateTasks.length === 0) {
            terminal.append([
                chalk.bold("No updates available") +
                    " for the given modules and version range",
            ]);
        } else {
            terminal.append([
                new Message("Found " + chalk.bold("%s update%s") + ".", [
                    updateTasks.length,
                    pluralize(updateTasks.length),
                ]),
                "",
            ]);
        }
    });
    updtr.on("batch-update/updating", event => {
        projector.display(
            cmdToLines(event.updateTasks.map(updatingLine), event.cmd)
        );
    });
    updtr.on("batch-update/testing", event => {
        projector.display(
            cmdToLines(event.updateTasks.map(testingLine), event.cmd)
        );
    });
    updtr.on("batch-update/rollback", event => {
        projector.display(
            cmdToLines(event.updateTasks.map(rollbackLine), event.cmd)
        );
    });
    updtr.on("batch-update/result", event => {
        projector.stop();
        if (event.success === true) {
            terminal.append(
                event.updateTasks.map(event.success ? successLine : failLine)
            );
        }
        // Not showing the test stdout here when there was an error because
        // we will proceed with the sequential update.
    });
    updtr.on("sequential-update/updating", event => {
        projector.display(cmdToLines(updatingLine(event), event.cmd));
    });
    updtr.on("sequential-update/testing", event => {
        projector.display(cmdToLines(testingLine(event), event.cmd));
    });
    updtr.on("sequential-update/rollback", event => {
        projector.display(cmdToLines(rollbackLine(event), event.cmd));
    });
    updtr.on("sequential-update/result", event => {
        projector.stop();
        terminal.append([(event.success ? successLine : failLine)(event)]);
        if (reporterConfig.testStdout === true && event.success === false) {
            terminal.append([event.stdout]);
        }
    });
    updtr.on("end", ({ results }) => {
        const duration = msToString(Date.now() - startTime);
        const successful = filterSuccessfulUpdates(results);
        const failed = filterFailedUpdates(results);

        terminal.append([""]);

        if (successful.length > 0) {
            terminal.append([
                new Message(chalk.bold("%s successful") + " update%s.", [
                    successful.length,
                    pluralize(successful.length),
                ]),
            ]);
        }
        if (failed.length > 0) {
            terminal.append([
                new Message(chalk.bold("%s failed") + " update%s.", [
                    failed.length,
                    pluralize(failed.length),
                ]),
            ]);
        }
        if (excludedModules.length > 0) {
            const list = excludedModules.map(excludedLine);

            if (successful.length > 0 || failed.length > 0) {
                terminal.append([""]);
            }
            terminal.append(
                [
                    new Message(chalk.bold("%s skipped") + " module%s:", [
                        excludedModules.length,
                        pluralize(excludedModules.length),
                    ]),
                    "",
                ].concat(list)
            );
        }

        terminal.append(["", new Message("Finished after %s.", [duration])]);
    });
    updtr.on("error", err => void handleError(err));
}
