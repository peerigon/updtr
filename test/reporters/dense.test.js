import EventEmitter from "events";
import { WritableStreamBuffer } from "stream-buffers";
import unicons from "unicons";
import dense from "../../src/reporters/dense";
import Spinner from "../../src/reporters/util/Spinner";
import events from "../fixtures/events";

function setup(reporterConfig = {}) {
    const updtr = new EventEmitter();
    const stdout = new WritableStreamBuffer();

    stdout.isTTY = true;
    stdout.columns = 80;

    reporterConfig.stream = stdout;
    dense(updtr, reporterConfig);

    return {
        updtr,
        stdout,
    };
}

beforeAll(() => {
    // We need to replace platform-dependent characters with static counter parts to make snapshot testing
    // consistent across platforms.
    Spinner.prototype.valueOf = () => "...";
    unicons.cli = sign => {
        switch (sign) {
            case "circle":
                return "-";
            default:
                return "";
        }
    };
});

describe("dense()", () => {
    Object.keys(events).forEach(caseName => {
        describe(caseName, () => {
            it("should print the expected lines", async () => {
                const testCase = events[caseName];
                const { updtr, stdout } = setup(testCase.reporterConfig);

                await testCase.events.reduce(async (previous, [
                    eventName,
                    event,
                ]) => {
                    await previous;
                    updtr.emit(eventName, event);

                    // Faking async events
                    return Promise.resolve();
                }, Promise.resolve());

                const output = stdout.getContentsAsString("utf8");

                expect(
                    output.replace(
                        // We need to replace the timing because that is non-deterministic
                        /Finished after \d+\.\ds/,
                        "Finished after 1.0s"
                    )
                ).toMatchSnapshot(caseName);
            });
        });
    });
});
