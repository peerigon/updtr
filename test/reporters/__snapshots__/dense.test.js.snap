// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batch-update fail and show test stdout 1`] = `
"[?25l[JInstalling missing dependencies[90m...[39m
[90m> npm install [39m
...
[3A[JLooking for outdated modules[90m...[39m
[90m> npm outdated [39m
...
[3A[JFound [1m4 updates[22m.

[J[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.0.1[90m...[39m
[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.1.0[90m...[39m
[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 1.0.0[90m...[39m
[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.0.1[90m...[39m
[90m> npm install [39m
...
[6A[J[33m-[39m [1mmodule[22m [90mtesting...[39m
[33m-[39m [1mmodule[22m [90mtesting...[39m
[33m-[39m [1mmodule[22m [90mtesting...[39m
[33m-[39m [1mmodule[22m [90mtesting...[39m
[90m> npm test [39m
...
[6A[J[31m-[39m [1m[31mmodule[39m[22m [90mrolling back[39m 0.0.1 [90mâ†’[39m 0.0.0[90m...[39m
[31m-[39m [1m[31mmodule[39m[22m [90mrolling back[39m 0.1.0 [90mâ†’[39m 0.0.0[90m...[39m
[31m-[39m [1m[31mmodule[39m[22m [90mrolling back[39m 1.0.0 [90mâ†’[39m 0.0.0[90m...[39m
[31m-[39m [1m[31mmodule[39m[22m [90mrolling back[39m 0.0.1 [90mâ†’[39m 0.0.0[90m...[39m
[90m> npm install [39m
...
[6A[J
[J[1m2 successful[22m updates.
[J[1m1 failed[22m update.
[J
Finished after 1.0s.
"
`;

exports[`batch-update success and show test stdout 1`] = `
"[?25l[JInstalling missing dependencies[90m...[39m
[90m> npm install [39m
...
[3A[JLooking for outdated modules[90m...[39m
[90m> npm outdated [39m
...
[3A[JFound [1m4 updates[22m.

[J[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.0.1[90m...[39m
[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.1.0[90m...[39m
[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 1.0.0[90m...[39m
[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.0.1[90m...[39m
[90m> npm install [39m
...
[6A[J[33m-[39m [1mmodule[22m [90mtesting...[39m
[33m-[39m [1mmodule[22m [90mtesting...[39m
[33m-[39m [1mmodule[22m [90mtesting...[39m
[33m-[39m [1mmodule[22m [90mtesting...[39m
[90m> npm test [39m
...
[6A[J[32m-[39m [1mmodule[22m 0.0.1 [90msuccess[39m
[32m-[39m [1mmodule[22m 0.1.0 [90msuccess[39m
[32m-[39m [1mmodule[22m 1.0.0 [90msuccess[39m
[32m-[39m [1mmodule[22m 0.0.1 [90msuccess[39m
[J
[J[1m3 successful[22m updates.
[J
Finished after 1.0s.
"
`;

exports[`custom config and only excluded modules 1`] = `
"[?25l[JRunning updtr with custom configuration:

- exclude: b, c

[JInstalling missing dependencies[90m...[39m
[90m> npm install [39m
...
[3A[JLooking for outdated modules[90m...[39m
[90m> npm outdated [39m
...
[3A[J[1mNo updates available[22m for the given modules and version range
[J
[J[1m3 skipped[22m modules:

[90m-[39m [1ma[22m [90mgit[39m
[90m-[39m [1mb[22m [90mexcluded[39m
[90m-[39m [1mc[22m [90mexcluded[39m
[J
Finished after 1.0s.
"
`;

exports[`custom config and sequential-update with mixed success and show test stdout 1`] = `
"[?25l[JRunning updtr with custom configuration:

- exclude: b, c

[JInstalling missing dependencies[90m...[39m
[90m> npm install [39m
...
[3A[JLooking for outdated modules[90m...[39m
[90m> npm outdated [39m
...
[3A[JFound [1m4 updates[22m.

[J[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.0.1[90m...[39m
[90m> npm install [39m
...
[3A[J[33m-[39m [1mmodule[22m [90mtesting...[39m
[90m> npm test [39m
...
[3A[J[32m-[39m [1mmodule[22m 0.0.1 [90msuccess[39m
[J[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.1.0[90m...[39m
[90m> npm install [39m
...
[3A[J[33m-[39m [1mmodule[22m [90mtesting...[39m
[90m> npm test [39m
...
[3A[J[31m-[39m [1m[31mmodule[39m[22m [90mrolling back[39m 0.1.0 [90mâ†’[39m 0.0.0[90m...[39m
[90m> npm install [39m
...
[3A[J[31m-[39m [1m[31mmodule[39m[22m 0.1.0 [90mfailed[39m
[JThis is the test stdout
[J[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 1.0.0[90m...[39m
[90m> npm install [39m
...
[3A[J[33m-[39m [1mmodule[22m [90mtesting...[39m
[90m> npm test [39m
...
[3A[J[32m-[39m [1mmodule[22m 1.0.0 [90msuccess[39m
[J[33m-[39m [1mmodule[22m [90mupdating[39m 0.0.0 [90mâ†’[39m 0.0.1[90m...[39m
[90m> npm install [39m
...
[3A[J[33m-[39m [1mmodule[22m [90mtesting...[39m
[90m> npm test [39m
...
[3A[J[31m-[39m [1m[31mmodule[39m[22m [90mrolling back[39m 0.0.1 [90mâ†’[39m 0.0.0[90m...[39m
[90m> npm install [39m
...
[3A[J[31m-[39m [1m[31mmodule[39m[22m 0.0.1 [90mfailed[39m
[JThis is the test stdout
[J
[J[1m2 successful[22m updates.
[J[1m2 failed[22m updates.
[J
Finished after 1.0s.
"
`;

exports[`no outdated modules 1`] = `
"[?25l[JInstalling missing dependencies[90m...[39m
[90m> npm install [39m
...
[3A[JLooking for outdated modules[90m...[39m
[90m> npm outdated [39m
...
[3A[JEverything [1mup-to-date[22m
[J
[J
Finished after 1.0s.
"
`;
